File: ProjectCS/UI/FacultyPage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SessionResultsCS.UI
{
    public class FacultyPage : LoopPage
    {
        private Faculty _faculty;

        public FacultyPage(Faculty faculty)
        {
            _faculty = faculty;
            SetPicker(new MenuActionPicker([
                ("Перейти к специальности", GoToSpeciality),
                ("Изменить название факультета", ChangeFacultyName),
                ("Добавить специальность", AddSpeciality),
                ("Удалить специальность", DeleteSpeciality),
                ("Сохранить изменения и вернуться в главное меню", StopLoop),
            ]));
        }

        protected override void Loop()
        {
            Console.WriteLine($"======== {_faculty} ========");
            Console.WriteLine("Специальности:");
            PrintSpecialities();
        }


        private void GoToSpeciality()
        {
            int selectedIndex = SelectSpeciality();
            if (selectedIndex != -1)
            {
                new SpecialityPage(_faculty.GetSpecialityAt(selectedIndex)).Start();
            }
        }

        private void ChangeFacultyName()
        {
            try
            {
                Console.Write("Введите новое название факультета: ");
                string name = Console.ReadLine()!;
                _faculty.Name = name;
            }
            catch (ArgumentException)
            {
                Console.WriteLine("Введено некорректное название");
            }
        }

        private void AddSpeciality()
        {
            Console.Write("Введите код новой специальности: ");
            string code = Console.ReadLine()!;
            Console.Write("Введите название новой специальности: ");
            string name = Console.ReadLine()!;
            
            try
            {
                Speciality speciality = new Speciality(_faculty, code, name);
                _faculty.AddSpeciality(speciality);
            }
            catch (ArgumentException)
            {
                Console.WriteLine("Не удалось добавить специальность");
            }
        }

        private void DeleteSpeciality()
        {
            int selectedIndex = SelectSpeciality();
            if (selectedIndex != -1)
            {
                _faculty.DeleteSpeciality(_faculty.GetSpecialityAt(selectedIndex));
            }
        }


        private void PrintSpecialities()
        {
            foreach (var (index, speciality) in _faculty.GetSpecialities().Index())
            {
                Console.WriteLine($"{index + 1}. {speciality}");
            }
        }

        private int SelectSpeciality()
        {
            Console.Write("Введите номер специальности: ");
            string input = Console.ReadLine()!;
            if (int.TryParse(input, out int value) && (value >= 1 && value <= _faculty.GetSpecialityCount()))
            {
                return value - 1;
            }
            else
            {
                Console.WriteLine("Введён неверный номер");
                return -1;
            }
        }
    }
}

File: Project/UI/FacultyPage.cpp
#include "UI/FacultyPage.hpp"

#include <format>
#include <iostream>


FacultyPage::FacultyPage(std::shared_ptr<Faculty> faculty) {
	m_faculty = faculty;
	SetPicker(std::make_unique<MenuActionPicker>(std::vector<MenuActionPicker::Item>{
		{ "Перейти к специальности", [this]() { GoToSpeciality(); } }, 
		{ "Изменить название факультета", [this]() { ChangeFacultyName(); } },
		{ "Добавить специальность", [this]() { AddSpeciality(); } },
		{ "Удалить специальность", [this]() { DeleteSpeciality(); } },
		{ "Сохранить изменения и вернуться в главное меню", [this]() { StopLoop(); } }
	}));
}

void FacultyPage::Loop() {
	std::cout << std::format("======== {} ========\n", m_faculty->GetName());
	std::cout << "Специальности:\n";
	PrintSpecialities();
}

void FacultyPage::GoToSpeciality() {
	int selectedIndex = SelectSpeciality();
	if (selectedIndex != -1) {
		SpecialityPage(m_faculty->GetSpecialityAt(selectedIndex)).Start();
	}
}

void FacultyPage::ChangeFacultyName() {
	try {
		std::cout << "Введите новое название факультета: ";
		std::string name;
		std::getline(std::cin, name);
		m_faculty->SetName(name);
	} catch (std::invalid_argument) {
		std::cout << "Введено некорректное название\n";
	}
}

void FacultyPage::AddSpeciality() {
	std::cout << "Введите код новой специальности: ";
	std::string code;
	std::getline(std::cin, code);

	std::cout << "Введите название новой специальности: ";
	std::string name;
	std::getline(std::cin, name);

	try {
		m_faculty->AddSpeciality(std::make_shared<Speciality>(code, name));
	} catch (std::invalid_argument) {
		std::cout << "Не удалось добавить специальность\n";
	}
}

void FacultyPage::DeleteSpeciality() {
	int selectedIndex = SelectSpeciality();
	if (selectedIndex != -1) {
		m_faculty->DeleteSpeciality(m_faculty->GetSpecialityAt(selectedIndex));
	}
}

void FacultyPage::PrintSpecialities() {
	for (int i = 0; i < m_faculty->GetSpecialityCount(); i++) {
		std::cout << std::format("{}. {}\n", i + 1, m_faculty->GetSpecialityAt(i)->ToString());
	}
}

int FacultyPage::SelectSpeciality() {
	int value;
	std::cout << "Введите номер специальности: ";
	if ((std::cin >> value) && (value >= 1 && value <= m_faculty->GetSpecialityCount())) {
		std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
		return value - 1;
	} else {
		std::cin.clear();
		std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
		return -1;
	}
}
